aioxmpp/bookmarks/service.py

```diff
@@ -290,58 +303,107 @@ def sync(self):
         yield from self.get_bookmarks()
 
     @asyncio.coroutine
-    def add_bookmark(self, new_bookmark):
+    def add_bookmark(self, new_bookmark, max_retries=3):
         """
-        Add a bookmark.
+        Add a bookmark and check whether it was successfully added to the
+        bookmark list. Already existant bookmarks are not added twice.
 
-        Already existant bookmarks are not added twice.
+        :raises RuntimeError: if the bookmark is not in the bookmark list
+                              after `max_retries` retries.
         """
-        bookmarks = yield from self._get_bookmarks()
-        for bookmark in bookmarks:
-            if bookmark == new_bookmark:
-                break
-        else:
-            bookmarks.append(new_bookmark)
-        yield from self._set_bookmarks(bookmarks)
+        with (yield from self._lock):
+            bookmarks = yield from self._get_bookmarks()
+
+            if new_bookmark not in bookmarks:
+                bookmarks.append(new_bookmark)
+            yield from self._set_bookmarks(bookmarks)
+
+            retries = 0
+            bookmarks = yield from self._get_bookmarks()
+            while retries < max_retries:
+                if new_bookmark in bookmarks:
+                    break
+                bookmarks.append(new_bookmark)
+                yield from self._set_bookmarks(bookmarks)
+                bookmarks = yield from self._get_bookmarks()
+                retries += 1
+
+            self._diff_emit_update(bookmarks)
 
-        self._diff_emit_update(bookmarks)
+            if new_bookmark not in bookmarks:
+                raise RuntimeError("Could not add bookmark")
 
     @asyncio.coroutine
-    def remove_bookmark(self, bookmark_to_remove):
+    def remove_bookmark(self, bookmark_to_remove, max_retries=3):
```

Comment:
Yes, I already considered that. But I think that otherwise the semantics are correct (removing *one* bookmark if there are several equal ones and not raising, since there is no reason to complain if we want to remove a bookmark and it's already gone).